### PR TITLE
Improve the error output for conflicting params and provide better option help describing environment variables used

### LINKED ISSUE #423: Better explanations for env vars that are being used
We need a better explanation for what environment variables `rsconnect-python` checks. 

For example if you're deploying to a server with `-s` + `-k` but also have that server saved, we throw this warning but it's unclear why:

```
Connect detected CLI commands and/or environment variables that overlap with stored credential.
Check your environment variables (e.g. CONNECT_API_KEY) to make sure you want them to be used.
Credential paremeters are taken with the following precedence: stored > CLI > environment.
To ignore an environment variable, override it in the CLI with an empty string (e.g. -k '').
```

Additionally, I had a `CONNECT_SERVER` environment variable and didn't realize that was being picked up by default and breaking deployments where `-n` was used. If `CONNECT_SERVER` is defined we try to use the `--server` argument in deployments but `-s` and `-n` cannot be used together.

example, if `CONNECT_SERVER` is defined:

```
# rsconnect deploy api -n localhost --title test-title ./bundles/python-bottle/
Error: -n/--name cannot be specified in conjunction with options -s/--server
```

I think it'd be also helpful if we included a way to ignore any of those set variables.

### PR DESCRIPTION
**Pull request by sagerb**
## Intent

Provide visibility into what environment variables are used as defaults for options. Also provide the source of where options were set (via environment variable, defaults, command line) when producing error messages to show conflicting values across options.

Resolves #423 

## Type of Change

- [ ] Bug Fix            
- [X] New Feature       
- [ ] Breaking Change   

## Approach

The click library provides an interface that will describe the source of an option. This source was added to the string being thrown when an exception was being produced for the multi-option validation errors.

To better enable the user to resolve the issue when they receive one of these errors, I also enhanced the help output for those options which also may be set via an environment variable.

Detailed description of what I've changed: 
- use click library to identify source of conflicting parameters
- update help output to display possible source of options - env or cmd
- internal documentation of failure case where complex logic used
- Types added for functions where click context passed down
- Type errors fixed after new type hints
- Errors updated to mention environment variables
- Unit tests updated
- Most complex commands now output what parameters they are using (values and their source), when verbose option provided

**Examples of new help output:**

2023-10-24 at 10 24 AM

**Examples of enhanced error output**

2023-11-09 at 2 12 PM

## Automated Tests

Some failure conditions are already included within the unit tests.

I've outlined the combinations of conflicting or missing options below (and within the code) that will produce validation errors. After discussing with @kgartland-rstudio , he will be adding validation for the messages below to rsconnect-python tests present within the Connect repo.

For example: rsconnect deploy api --name localhost ./python-bottle-py3

    should fail w/
    -s/--server or CONNECT_SERVER
    -T/--token or SHINYAPPS_TOKEN or RSCLOUD_TOKEN
    -S/--secret or SHINYAPPS_SECRET or RSCLOUD_SECRET
    -A/--account or SHINYAPPS_ACCOUNT

    FAILURE if not any of:
    -n/--name
    -s/--server or CONNECT_SERVER
    -T/--token or SHINYAPPS_TOKEN or RSCLOUD_TOKEN
    -S/--secret or SHINYAPPS_SECRET or RSCLOUD_SECRET
    -A/--account or SHINYAPPS_ACCOUNT

    FAILURE if any of:
    -k/--api-key or CONNECT_API_KEY
    -i/--insecure or CONNECT_INSECURE
    -c/--cacert or CONNECT_CA_CERTIFICATE
    AND any of:
    -T/--token or SHINYAPPS_TOKEN or RSCLOUD_TOKEN
    -S/--secret or SHINYAPPS_SECRET or RSCLOUD_SECRET
    -A/--account or SHINYAPPS_ACCOUNT

    FAILURE if specify -s/--server or CONNECT_SERVER and it includes "posit.cloud" or "rstudio.cloud"
    and not specified all of following:
    -T/--token or SHINYAPPS_TOKEN or RSCLOUD_TOKEN
    -S/--secret or SHINYAPPS_SECRET or RSCLOUD_SECRET

    FAILURE if any of following are specified, without the rest:
    -T/--token or SHINYAPPS_TOKEN or RSCLOUD_TOKEN
    -S/--secret or SHINYAPPS_SECRET or RSCLOUD_SECRET
    -A/--account or SHINYAPPS_ACCOUNT

**Impacted commands needing validation:**

- rsconnect system caches delete
- rsconnect system caches list
- rsconnect content build run
- rsconnect content build logs
- rsconnect content build history
- rsconnect content build ls
- rsconnect content build rm
- rsconnect content build add
- rsconnect content download-bundle
- rsconnect content describe
- rsconnect content search
- rsconnect write-manifest api
- rsconnect write-manifest bokeh
- rsconnect write-manifest dash
- rsconnect write-manifest fastapi
- rsconnect write-manifest flask
- rsconnect write-manifest notebook
- rsconnect write-manifest quarto
- rsconnect write-manifest shiny
- rsconnect write-manifest streamlit
- rsconnect write-manifest voila
- rsconnect deploy api 
- rsconnect deploy bokeh
- rsconnect deploy dash
- rsconnect deploy fastapi
- rsconnect deploy flask
- rsconnect deploy html
- rsconnect deploy manifest
- rsconnect deploy notebook
- rsconnect deploy other-content
- rsconnect deploy quarto
- rsconnect deploy shiny
- rsconnect deploy streamlit
- rsconnect deploy voila
- rsconnect remove
- rsconnect details
- rsconnect add

**help output to include env variables - static string**

Any check of help against a single command which accepts server parameters will validate for all usage. 

For example: `rsconnect deploy bokeh --help`:

```
Usage: rsconnect deploy bokeh [OPTIONS] DIRECTORY [EXTRA_FILES]...

  Deploy a Bokeh Application module to Posit Connect, Posit Cloud, or
  shinyapps.io (if supported by the platform). The "directory" argument must
  refer to an existing directory that contains the application code.

Options:
  -n, --name TEXT                 The nickname of the Posit Connect server to
                                  deploy to.
  -s, --server TEXT               The URL for the Posit Connect server to
                                  deploy to. (Also settable via CONNECT_SERVER
                                  environment variable.)
  -k, --api-key TEXT              The API key to use to authenticate with
                                  Posit Connect. (Also settable via
                                  CONNECT_API_KEY environment variable.)
  -i, --insecure                  Disable TLS certification/host validation.
                                  (Also settable via CONNECT_INSECURE
                                  environment variable.)
  -c, --cacert FILE               The path to trusted TLS CA certificates.
                                  (Also settable via CONNECT_CA_CERTIFICATE
                                  environment variable.)
  -v, --verbose                   Enable verbose output. Use -vv for very
                                  verbose (debug) output.
  -N, --new                       Force a new deployment, even if there is
                                  saved metadata from a previous deployment.
                                  Cannot be used with --app-id.
  -a, --app-id TEXT               Existing app ID or GUID to replace. Cannot
                                  be used with --new.
  -t, --title TEXT                Title of the content (default is the same as
                                  the filename).
  -E, --environment TEXT          Set an environment variable. Specify a value
                                  with NAME=VALUE, or just NAME to use the
                                  value from the local environment. May be
                                  specified multiple times. [v1.8.6+]
  --no-verify                     Don't access the deployed content to verify
                                  that it started correctly.
  -A, --account TEXT              The shinyapps.io/Posit Cloud account name.
                                  (Also settable via SHINYAPPS_ACCOUNT
                                  environment variable.)
  -T, --token TEXT                The shinyapps.io/Posit Cloud token. (Also
                                  settable via SHINYAPPS_TOKEN or
                                  RSCLOUD_TOKEN environment variables.)
  -S, --secret TEXT               The shinyapps.io/Posit Cloud token secret.
                                  (Also settable via SHINYAPPS_SECRET or
                                  RSCLOUD_SECRET environment variables.)
  -I, --image TEXT                Target image to be used during content build
                                  and execution. This option is only
                                  applicable if the Connect server is
                                  configured to use off-host execution.
  --disable-env-management        Shorthand to disable environment management
                                  for both Python and R.
  --disable-env-management-py     Disable Python environment management for
                                  this bundle. Connect will not create an
                                  environment or install packages. An
                                  administrator must install the required
                                  packages in the correct Python environment
                                  on the Connect server.
  --disable-env-management-r      Disable R environment management for this
                                  bundle. Connect will not create an
                                  environment or install packages. An
                                  administrator must install the required
                                  packages in the correct R environment on the
                                  Connect server.
  -e, --entrypoint TEXT           The module and executable object which
                                  serves as the entry point for the Bokeh
                                  Application (defaults to app)
  -x, --exclude TEXT              Specify a glob pattern for ignoring files
                                  when building the bundle. Note that your
                                  shell may try to expand this which will not
                                  do what you expect. Generally, it's safest
                                  to quote the pattern. This option may be
                                  repeated.
  -p, --python PATH               Path to Python interpreter whose environment
                                  should be used. The Python environment must
                                  have the rsconnect package installed.
  -g, --force-generate            Force generating "requirements.txt", even if
                                  it already exists.
  -V, --visibility [public|private]
                                  The visibility of the resource being
                                  deployed. (shinyapps.io only; must be public
                                  (default) or private)
  --help                          Show this message and exit.
```

## Directions for Reviewers

You can validate some of the combinations above manually, as well as execute the help commands.

## Checklist

- [x] I have updated [CHANGELOG.md](../CHANGELOG.md) to cover notable changes.
- [x] I have updated all related GitHub issues to reflect their current state.

### CODE REVIEW

**Comment by kgartland-rstudio** [comment]
I can't add a server on this branch.

rsconnect version: 1.20.1.dev111+g49ea42c

```
> rsconnect add -s http://localhost:3939 -k $DEFAULT_ADMIN_API_KEY -n ci-test
Detected the following inputs:
    name: COMMANDLINE
    server: COMMANDLINE
    api_key: COMMANDLINE
    insecure: DEFAULT
Traceback (most recent call last):
  File "/Users/kgartland/.pyenv/versions/3.11.6/bin/rsconnect", line 8, in <module>
    sys.exit(cli())
             ^^^^^
  File "/Users/kgartland/.pyenv/versions/3.11.6/lib/python3.11/site-packages/click/core.py", line 1157, in __call__
    return self.main(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/kgartland/.pyenv/versions/3.11.6/lib/python3.11/site-packages/click/core.py", line 1078, in main
    rv = self.invoke(ctx)
         ^^^^^^^^^^^^^^^^
  File "/Users/kgartland/.pyenv/versions/3.11.6/lib/python3.11/site-packages/click/core.py", line 1688, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/kgartland/.pyenv/versions/3.11.6/lib/python3.11/site-packages/click/core.py", line 1434, in invoke
    return ctx.invoke(self.callback, **ctx.params)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/kgartland/.pyenv/versions/3.11.6/lib/python3.11/site-packages/click/core.py", line 783, in invoke
    return __callback(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/kgartland/.pyenv/versions/3.11.6/lib/python3.11/site-packages/rsconnect/main.py", line 175, in wrapper
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/Users/kgartland/.pyenv/versions/3.11.6/lib/python3.11/site-packages/click/decorators.py", line 33, in new_func
    return f(get_current_context(), *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/kgartland/.pyenv/versions/3.11.6/lib/python3.11/site-packages/rsconnect/main.py", line 515, in add
    validation.validate_connection_options(
  File "/Users/kgartland/.pyenv/versions/3.11.6/lib/python3.11/site-packages/rsconnect/validation.py", line 97, in validate_connection_options
    present_connect_options = _get_present_options(connect_options, ctx)
                              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/kgartland/.pyenv/versions/3.11.6/lib/python3.11/site-packages/rsconnect/validation.py", line 17, in _get_present_options
    source = ctx.get_parameter_source(parts[1]).name  # type: ignore
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AttributeError: 'NoneType' object has no attribute 'name'
```

File: rsconnect/validation.py [hunk 1]
```diff
@@ -1,46 +1,119 @@
...
+def _get_present_options(
+    options: typing.Dict[str, typing.Optional[typing.Any]],
+    ctx: click.Context = None,
+) -> typing.List[str]:
+    result: typing.List[str] = []
+    for k, v in options.items():
+        if v:
+            parts = k.split("--")
+            if ctx and len(parts) == 2:
+                source = ctx.get_parameter_source(parts[1]).name  # type: ignore
```
Comment (tdstein) [inline rsconnect/validation.py:17]: It looks like ctx.get_parameter_source(parts[1]) returns None in some cases according to the @kgartland-rstudio 's bug report.

**Comment by mmarchetti** [comment]
The CI failures here are because we are using `types-click`; see the [note here](https://pypi.org/project/types-click/) about Click 8.0 already including types.

File: rsconnect/validation.py [hunk 2]
```diff
@@ -1,46 +1,121 @@
...
+    options: typing.Dict[str, typing.Optional[typing.Any]],
+    ctx: click.Context,
+) -> typing.List[str]:
+    result: typing.List[str] = []
+    for k, v in options.items():
+        if v:
+            parts = k.split("--")
+            if ctx and len(parts) == 2:
+                varName = parts[1].replace("-", "_")
+                source = ctx.get_parameter_source(varName)  # type: ignore
```
Comment (mmarchetti) [inline rsconnect/validation.py:18]: This function was introduced in Click 8.0. Now that our minimum python version is 3.7, we can require click >= 8.0.

**Comment by kgartland-rstudio** [comment]
Looks like I can't deploy on the latest build:

```
> rsconnect version
1.20.1.dev113+g2cdc988
```

```
> rsconnect deploy api bundles/python-bottle/ -s https://rsc.radixu.com -k $DOGFOOD_API_KEY
Detected the following inputs:
    server:	https://rsc.radixu.com (from COMMANDLINE)
    api_key:	********** (from COMMANDLINE)
    insecure:	False (from DEFAULT)
    exclude:	() (from DEFAULT)
    new:	False (from DEFAULT)
    force_generate:	False (from DEFAULT)
    directory:	bundles/python-bottle/ (from COMMANDLINE)
    extra_files:	() (from COMMANDLINE)
    env_vars:	{} (from DEFAULT)
    no_verify:	False (from DEFAULT)
    app_mode:	python-api
Connect detected CLI commands and/or environment variables that overlap with stored credential.
Check your environment variables (e.g. CONNECT_API_KEY) to make sure you want them to be used.
Credential parameters are taken with the following precedence: stored > CLI > environment.
To ignore an environment variable, override it in the CLI with an empty string (e.g. -k '').
Traceback (most recent call last):
  File "/Users/kgartland/.pyenv/versions/3.12.0/lib/python3.12/site-packages/rsconnect/main.py", line 96, in wrapper
    result = func(*args, **kwargs)
             ^^^^^^^^^^^^^^^^^^^^^
  File "/Users/kgartland/.pyenv/versions/3.12.0/lib/python3.12/site-packages/click/decorators.py", line 33, in new_func
    return f(get_current_context(), *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/kgartland/.pyenv/versions/3.12.0/lib/python3.12/site-packages/rsconnect/main.py", line 1432, in deploy_app
    ce = RSConnectExecutor(**kwargs)
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/kgartland/.pyenv/versions/3.12.0/lib/python3.12/site-packages/rsconnect/api.py", line 412, in __init__
    self.setup_remote_server(
  File "/Users/kgartland/.pyenv/versions/3.12.0/lib/python3.12/site-packages/rsconnect/api.py", line 506, in setup_remote_server
    header_output = self.output_overlap_details("api-key", header_output)
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/kgartland/.pyenv/versions/3.12.0/lib/python3.12/site-packages/rsconnect/api.py", line 469, in output_overlap_details
    self.logger.warning(f"stored {cli_param} value overrides the {cli_param} value from {source.name}")
                                                                                         ^^^^^^^^^^^
AttributeError: 'NoneType' object has no attribute 'name'
Internal error: 'NoneType' object has no attribute 'name'
```

**Comment by kgartland-rstudio** [comment]
We deploying content where a filename is required (rather than just the path to an api for example, there's a `kwargs` input that's detected and I'm unsure if that's intended.  Here's an example:

```
> rsconnect deploy html index.html -n dogfood
Detected the following inputs:
    path:	index.html (from COMMANDLINE)
    extra_files:	() (from COMMANDLINE)
    exclude:	() (from DEFAULT)
    env_vars:	{} (from DEFAULT)
    new:	False (from DEFAULT)
    name:	dogfood (from COMMANDLINE)
    insecure:	False (from DEFAULT)
    no_verify:	False (from DEFAULT)
    kwargs:	{'ctx': <click.core.Context object at 0x105956210>, 'connect_server': None, 'path': 'index.html', 'entrypoint': None, 'extra_files': (), 'exclude': (), 'title': None, 'env_vars': {}, 'verbose': 0, 'new': False, 'app_id': None, 'name': 'dogfood', 'server': None, 'api_key': None, 'insecure': False, 'cacert': None, 'account': None, 'token': None, 'secret': None, 'no_verify': False, 'kwargs': {...}} (from <source unknown>)
```

You can replicate this by deploying an html doc as above or a notebook:

```
> rsconnect deploy notebook ./stock-report-jupyter.ipynb quandl-wiki-tsla.json.gz -n dogfood
Detected the following inputs:
    name:	dogfood (from COMMANDLINE)
    insecure:	False (from DEFAULT)
    static:	False (from DEFAULT)
    new:	False (from DEFAULT)
    force_generate:	False (from DEFAULT)
    file:	./stock-report-jupyter.ipynb (from COMMANDLINE)
    extra_files:	('quandl-wiki-tsla.json.gz',) (from COMMANDLINE)
    hide_all_input:	False (from DEFAULT)
    hide_tagged_input:	False (from DEFAULT)
    env_vars:	{} (from DEFAULT)
    no_verify:	False (from DEFAULT)
    kwargs:	{'ctx': <click.core.Context object at 0x103025510>, 'name': 'dogfood', 'server': None, 'api_key': None, 'insecure': False, 'cacert': None, 'static': False, 'new': False, 'app_id': None, 'title': None, 'python': None, 'force_generate': False, 'verbose': 0, 'file': './stock-report-jupyter.ipynb', 'extra_files': ('quandl-wiki-tsla.json.gz',), 'hide_all_input': False, 'hide_tagged_input': False, 'env_vars': {}, 'image': None, 'disable_env_management': None, 'env_management_py': None, 'env_management_r': None, 'no_verify': False, 'kwargs': {...}} (from <source unknown>)
```

**Comment by kgartland-rstudio** [comment]
The output of the rsconnect content search subcommands is no longer valid json due to printing out the inputs:

```
> rsconnect content search --title-contains "binary python-falcon-asgi" --server http://localhost:3939 --api-key ${DEFAULT_ADMIN_API_KEY}
Detected the following inputs:
    server:	http://localhost:3939 (from COMMANDLINE)
    api_key:	********** (from COMMANDLINE)
    insecure:	False (from DEFAULT)
    published:	False (from DEFAULT)
    unpublished:	False (from DEFAULT)
    content_type:	() (from DEFAULT)
    title_contains:	binary python-falcon-asgi (from COMMANDLINE)
[
  {
    "guid": "4b01054f-2b9c-455b-b4c0-971d339a03b3",
    "name": "binary_python-falcon-asgi",
    "title": "binary python-falcon-asgi",
    "description": "",
    "access_type": "acl",
    "connection_timeout": null,
    "read_timeout": null,
    "init_timeout": null,
    "idle_timeout": null,
    "max_processes": null,
    "min_processes": null,
    "max_conns_per_process": null,
    "load_factor": null,
    "memory_request": null,
    "memory_limit": null,
    "cpu_request": null,
    "cpu_limit": null,
    "amd_gpu_limit": null,
    "nvidia_gpu_limit": null,
    "service_account_name": null,
    "default_image_name": null,
    "created_time": "2023-11-09T17:03:21Z",
    "last_deployed_time": "2023-11-09T17:03:21Z",
    "bundle_id": "3",
    "app_mode": "python-fastapi",
    "content_category": "",
    "parameterized": false,
    "cluster_name": "Local",
    "image_name": null,
    "r_version": null,
    "py_version": "3.11.3",
    "quarto_version": null,
    "r_environment_management": null,
    "default_r_environment_management": null,
    "py_environment_management": true,
    "default_py_environment_management": null,
    "run_as": null,
    "run_as_current_user": false,
    "owner_guid": "bce9013d-e755-4a44-abb5-45c6591bfac6",
    "content_url": "http://localhost:3939/content/4b01054f-2b9c-455b-b4c0-971d339a03b3/",
    "dashboard_url": "http://localhost:3939/connect/#/apps/4b01054f-2b9c-455b-b4c0-971d339a03b3",
    "app_role": "owner",
    "id": "3"
  }
]
```

While this just causes the rsconnect-python tests to fail, this makes is especailly hard to script against, which I think was the reason these commands were added initially. 

I think it'd be best for the `Detected the following inputs...` text to be exposed in the verbose mode only across all commands.

**Comment by sagerb** [comment]
Thanks @kgartland-rstudio . Those are good findings.

My suggestion is to suppress the kwargs argument when displaying input parameters and only output that list if the verbose flag is given. I’ll still output the enhanced explanation if there is an overlap, but otherwise, most of the commands should not see any difference, allowing them to not break existing scripts.

**Comment by kgartland-rstudio** [comment]
> Thanks @kgartland-rstudio . Those are good findings.
> 
> My suggestion is to suppress the kwargs argument when displaying input parameters and only output that list if the verbose flag is given. I’ll still output the enhanced explanation if there is an overlap, but otherwise, most of the commands should not see any difference, allowing them to not break existing scripts.

Sounds good, thanks!

**Review by mmarchetti — APPROVED** [review]

**Comment by kgartland-rstudio** [comment]
✅ Tests are passing against this branch with the new server tests:
https://github.com/rstudio/connect/actions/runs/6826092617

✅ Verified enviorment detection data is now only in `-v` mode
✅ Verified deploying single file/notebook no longer shows the `kwargs` output.

🚢 Merge away!

**Pull request merged**